akka-http-core/src/test/scala/akka/http/impl/engine/client/PrepareResponseSpec.scala

```diff
@@ -0,0 +1,243 @@
+      inSub.expectRequest(1)
+      inSub.sendNext(messageEnd)
+      entityProbe.expectComplete()
+
+      // and that demand should go downstream
+      // since the chunk end was consumed by the stage
+      inSub.expectRequest(1)
+
+    }
+
+    "not lose demand that comes in while handling strict entity" in {
+      implicit val mat = ActorMaterializer()
+
+      val inProbe = TestPublisher.manualProbe[ParserOutput.ResponseOutput]()
+      val responseProbe = TestSubscriber.manualProbe[HttpResponse]
+
+      Source.fromPublisher(inProbe)
+        .via(new PrepareResponse(parserSettings))
+        .to(Sink.fromSubscriber(responseProbe))
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .run()
+
+      val inSub = inProbe.expectSubscription()
+      val responseSub = responseProbe.expectSubscription()
+
+      responseSub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(chunkedStart)
+      val response = responseProbe.expectNext()
+
+      val entityProbe = TestSubscriber.manualProbe[ByteString]()
+      response.entity.dataBytes.to(Sink.fromSubscriber(entityProbe)).run()
+      val entitySub = entityProbe.expectSubscription()
+
+      entitySub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(chunk)
+      entityProbe.expectNext()
+
+      // now, before entity stream has completed
+      // there is upstream demand
+      responseSub.request(1)
+
+      // then chunk completes
+      entitySub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(messageEnd)
+      entityProbe.expectComplete()
+
+      // and that demand should go downstream
+      // since the chunk end was consumed by the stage
+      inSub.expectRequest(1)
+
+    }
+
+    "cancel entity stream when upstream cancels" in {
+      // not sure about this, but to document how it behaves
+      implicit val mat = ActorMaterializer()
+
+      val inProbe = TestPublisher.manualProbe[ParserOutput.ResponseOutput]()
+      val responseProbe = TestSubscriber.manualProbe[HttpResponse]
+
+      Source.fromPublisher(inProbe)
+        .via(new PrepareResponse(parserSettings))
+        .to(Sink.fromSubscriber(responseProbe))
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .run()
+
+      val inSub = inProbe.expectSubscription()
+      val responseSub = responseProbe.expectSubscription()
+
+      responseSub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(chunkedStart)
+      val response = responseProbe.expectNext()
+
+      val entityProbe = TestSubscriber.manualProbe[ByteString]()
+      response.entity.dataBytes.to(Sink.fromSubscriber(entityProbe)).run()
+      val entitySub = entityProbe.expectSubscription()
+
+      entitySub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(chunk)
+      entityProbe.expectNext()
+
+      // now before entity stream is completed,
+      // upstream cancels
+      responseSub.cancel()
+
+      inSub.expectCancellation()
+      entityProbe.expectComplete()
```

Comment:
This is not correct: the stage should ferry the remainder of this response through and then shut down.